lib/cache/memory.go

```diff
@@ -114,6 +116,16 @@ func (s *shard) compaction() {
 	s.lastCompaction = time.Now()
 }
 
+func (s *shard) shouldCompact() bool {
+	if s.compInterval == 0 {
+		return false
+	}
+	if time.Since(s.lastCompaction) < s.compInterval {
```

Comment:
If we aren't protecting `lastCompaction` with a lock then it'll need to be made into an atomic otherwise the write at the end of compaction will race.

Reply:
Nice find! 

After reading over the code a few times, the case where a race occurs should not be a problem as the compaction function should imediatly return within this section. (Under the safety of an rw lock)
```go
if shard.compOnRead && shard.shouldCompact() {
   shard.Lock()
   shard.compaction()
   shard.Unlock()
}
```
I am happy to remove `shard.shouldCompact()` from the `if` statement. The only downfall then is that we will always aquire a write lock on a read event in the case that `compact_on_read` is active. Keeping the current approach will 99% of the time only aquire a read lock and and for the 1% of the time aquire a write lock and imediatly return.

What do you think?

Reply:
I'm pretty sure it'll still race as there could be two concurrent operations where one is actively compacting and has the lock, and another enters `Get` and reads `lastCompaction` at the point where the other writes to it. Using an atomic would let you access `shouldCompact` lock free so the hot path would still be fast, might look at bit ugly though as you'd need to convert to/from `time.Time`.

Another alternative that (I think) would also solve your problem would be to check the TTL on items during `Get`, so if the item exists we also follow with:

```go
if !k.ts.IsZero() && time.Since(k.ts) >= m.ttl {
  return nil, types.ErrKeyNotFound
}
```

And therefore if it has expired but we haven't compacted it yet we still treat it as though it doesn't exist.

Reply:
Great sugestion! I went with the second approach as it keeps the code clean and simplifies the locking logic.